src/wlapplication.cc

```diff
@@ -1301,45 +1301,76 @@ void WLApplication::parse_commandline(int const argc, char const* const* const a
 	}
 }
 
+namespace {
+
+void throw_extra_value(const std::string& opt) {
+	throw ParameterError(
+	   CmdLineVerbosity::None, format(_("Command line parameter --%s can not use a value"), opt));
+}
+
+void throw_empty_value(const std::string& opt) {
+	throw ParameterError(
+	   CmdLineVerbosity::None, format(_("Empty value of command line parameter --%s"), opt));
+}
+
+void throw_exclusive(const std::string& opt, const std::string& other) {
+	throw ParameterError(
+	   CmdLineVerbosity::None,
+	   format(_("Command line parameters --%s and --%s can not be combined"), opt, other));
+}
+
+
+}  // namespace
+
+// Checks and returns whether `param` was set, but throws ParameterError if it also had a value.
+bool WLApplication::check_commandline_flag(const std::string& opt) {
+	auto found = commandline_.find(opt);
+	if (found == commandline_.end()) {
+		return false;
+	}
+	if (!found->second.empty()) {
+		throw_extra_value(opt);
+	}
+	commandline_.erase(found);
+	return true;
+}
+
+// Returns the value of `opt`. Only returns std::nullopt if `opt` was not used.
+// If `opt` was used without a value, then returns an empty string if `allow_empty` is true,
+// otherwise throws ParameterError.
+std::optional<std::string>
+WLApplication::get_commandline_option_value(const std::string& opt, const bool allow_empty) {
+	auto found = commandline_.find(opt);
+	if (found == commandline_.end()) {
+		return std::nullopt;
+	}
+	// need to copy before deletion for returning later
+	std::string rv = found->second;
+	if (!allow_empty && rv.empty()) {
+		throw_empty_value(opt);
+	}
+	commandline_.erase(found);
+	return rv;
+}
+
 /**
  * Parse the command line given in commandline_
  *
  * \throw a ParameterError if there were errors during parsing \e or if "--help"
  */
 void WLApplication::handle_commandline_parameters() {
-	auto throw_empty_value = [](const std::string& opt) {
-		throw ParameterError(
-		   CmdLineVerbosity::None, format(_("Empty value of command line parameter: %s"), opt));
-	};
 
-	auto throw_exclusive = [](const std::string& opt) {
-		throw ParameterError(
-		   CmdLineVerbosity::None, format(_("%s can not be combined with other actions"), opt));
-	};
-
-	if (commandline_.count("nosound") != 0u) {
-		SoundHandler::disable_backend();
-		commandline_.erase("nosound");
-	}
-	if (commandline_.count("verbose-i18n") != 0u) {
+	if (check_commandline_flag("verbose-i18n")) {
 		i18n::enable_verbose_i18n();
-		commandline_.erase("verbose-i18n");
-	}
-	if (commandline_.count("fail-on-lua-error") != 0u) {
-		g_fail_on_lua_error = true;
-		commandline_.erase("fail-on-lua-error");
-	}
-	if (commandline_.count("nozip") != 0u) {
-		set_config_bool("nozip", true);
-		commandline_.erase("nozip");
 	}
-	if (commandline_.count("localedir") != 0u) {
-		localedir_ = commandline_["localedir"];
-		commandline_.erase("localedir");
+
+	if (auto localedir_option = get_commandline_option_value("localedir");
```

Comment:
Instead of so many `auto` types that all resolve to the same type maybe define `using OptionalParameter = std::optional<std::string>`